src/schemata/date/DateFromIsoString.ts

```diff
@@ -0,0 +1,323 @@
+ * - Can be configured to require a time, time and timezone offset (e.g. `Z` or `±05:00`) or
+ *   neither (default is require both).
+ * - Dates may contain years, months, and days; years and months; or years
+ * - Times may contain hours, minutes, seconds, and milliseconds; hours, minutes, and
+ *   seconds; or hours and minutes.
+ * - Expanded years are permitted (e.g. `+002022` instead of `2022`)
+ *
+ * @since 1.0.0
+ */
+
+import { pipe } from 'fp-ts/function'
+import * as PB from '../../PatternBuilder'
+import { make, SchemaExt } from '../../SchemaExt'
+import * as D from '../../schemables/WithDate'
+import { match } from '../../internal/match'
+
+/**
+ * E.g. `2022`
+ *
+ * @since 1.0.0
+ * @category Pattern
+ */
+const year: PB.Pattern = pipe(PB.digit, PB.exactly(4))
+
+/**
+ * E.g. `+-002022`
+ *
+ * @since 1.0.0
+ * @category Pattern
+ */
+const expandedYear: PB.Pattern = pipe(
+  PB.characterClass(false, '+', '-'),
+  PB.subgroup,
+  PB.then(pipe(PB.digit, PB.exactly(6))),
+)
+
+/**
+ * E.g. 01, 02, 03, 04, 05, 06, 07, 08, 09, 10, 11, 12
+ *
+ * @since 1.0.0
+ * @category Pattern
+ */
+const month: PB.Pattern = pipe(
+  PB.char('0'),
+  PB.then(pipe(PB.integerRange(1, 9), PB.subgroup)),
+  PB.subgroup,
+  PB.or(pipe(PB.integerRange(10, 12), PB.subgroup)),
+  PB.subgroup,
+)
+
+/**
+ * E.g. 01, 02, 03, 04, 05, ..., 30, 31
+ *
+ * @since 1.0.0
+ * @category Pattern
+ */
+const day: PB.Pattern = pipe(
+  PB.char('0'),
+  PB.then(pipe(PB.integerRange(1, 9), PB.subgroup)),
+  PB.subgroup,
+  PB.or(pipe(PB.integerRange(10, 31), PB.subgroup)),
+  PB.subgroup,
+)
+
+/**
+ * E.g. `YYYY-MM-DD`
+ *
+ * @since 1.0.0
+ * @category Pattern
+ */
+const yearMonthDay: PB.Pattern = pipe(
+  year,
+  PB.or(expandedYear),
+  PB.subgroup,
+  PB.then(PB.char('-')),
+  PB.then(month),
+  PB.then(PB.char('-')),
+  PB.then(day),
+)
+
+/**
+ * E.g. `YYYY-MM`
+ *
+ * @since 1.0.0
+ * @category Pattern
+ */
+const yearMonth: PB.Pattern = pipe(
+  year,
+  PB.or(expandedYear),
+  PB.subgroup,
+  PB.then(PB.char('-')),
+  PB.then(month),
+)
+
+/**
+ * E.G. `YYYY-MM-DD` or `YYYY-MM` or `YYYY`
+ *
+ * @since 1.0.0
+ * @category Pattern
+ */
+const date: PB.Pattern = pipe(yearMonthDay, PB.or(yearMonth), PB.or(year), PB.subgroup)
+
+/**
+ * E.g. 00, 01, 02, 03, 04, 05, 06, 07, ..., 22, 23
+ *
+ * @since 1.0.0
+ * @category Pattern
+ */
+const hour: PB.Pattern = pipe(
+  PB.char('0'),
+  PB.then(pipe(PB.integerRange(0, 9), PB.subgroup)),
+  PB.subgroup,
+  PB.or(pipe(PB.integerRange(10, 23), PB.subgroup)),
+  PB.subgroup,
+)
+
+/**
+ * E.g. 00, 01, 02, 03, 04, 05, 06, 07, ..., 58, 59
+ *
+ * @since 1.0.0
+ * @category Pattern
+ */
+export const minutesSeconds: PB.Pattern = pipe(
+  PB.char('0'),
+  PB.then(pipe(PB.integerRange(0, 9), PB.subgroup)),
+  PB.subgroup,
+  PB.or(pipe(PB.integerRange(10, 59), PB.subgroup)),
+  PB.subgroup,
+)
+
+/**
+ * @since 1.0.0
+ * @category Pattern
+ */
+const milliseconds: PB.Pattern = pipe(PB.digit, PB.atLeastOne(), PB.subgroup)
```

Comment:
I think this must be exactly 3 digits

Reply:
I thought so too, but there is no requirement for exactly three digits.  You can try something like this in the browser or node:

```
2022-11-15T18:20:30.999999999999999-06:00
```

It should work